src/algebra/lie_algebra.lean

```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
 module endomorphisms.
 -/
-def of_endomorphism_algebra (M : Type v)
+instance of_endomorphism_algebra (M : Type v)
   [add_comm_group M] [module R M] : lie_algebra R (module.End R M) :=
-of_associative_algebra R (module.End R M)
+of_associative_algebra (module.End R M)
+
+/--
+The adjoint action of a Lie algebra on itself.
+-/
+def Ad (x : L) : L →ₗ[R] L :=
+{ to_fun := has_bracket.bracket x,
+  add    := by { intros, apply lie_add },
+  smul   := by { intros, apply lie_smul } }
+
+set_option class.instance_max_depth 35
+lemma is_morphism_Ad (x y : L) : (Ad ⁅x, y⁆ : L →ₗ[R] L) = ⁅Ad x, Ad y⁆ :=
+begin
+  ext z,
+  suffices : ⁅⁅x, y⁆, z⁆ = ⁅x, ⁅y, z⁆⁆ + ⁅⁅x, z⁆, y⁆, by {
+    dunfold of_endomorphism_algebra of_associative_algebra Ad
+            has_bracket.bracket ring_commutator.commutator,
+    simpa using this,
+  },
+  rw [eq_comm, ←lie_skew ⁅x, y⁆ z, ←lie_skew ⁅x, z⁆ y, ←lie_skew x z, lie_neg, neg_neg,
+      ←sub_eq_zero_iff_eq, sub_neg_eq_add, lie_ring.jacobi],
+end
+
+/--
+The bracket of a Lie algebra as a bilinear map.
+-/
+def bil_lie : L →ₗ[R] L →ₗ[R] L :=
+{ to_fun := Ad,
+  add    := by { unfold Ad, intros, ext, simp [add_lie], },
+  smul   := by { unfold Ad, intros, ext, simp, } }
+
+/--
+The bracket of a Lie algebra as a morphism of Lie algebras.
+-/
+def bil_lie_morphism : L →ₗ⁅R⁆ module.End R L := { bracket := is_morphism_Ad, ..bil_lie }
 
 end lie_algebra
 
+section lie_subalgebra
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+
+/--
+A Lie subalgebra of a Lie algebra is submodule that is closed under the Lie bracket.
+This is a sufficient condition for the subset itself to form a Lie algebra.
+-/
+structure lie_subalgebra extends submodule R L :=
+(bracket : ∀ {x y}, x ∈ carrier → y ∈ carrier → ⁅x, y⁆ ∈ carrier)
+
+instance lie_subalgebra_coe_submodule : has_coe (lie_subalgebra R L) (submodule R L) :=
+⟨lie_subalgebra.to_submodule⟩
+
+instance lie_subalgebra_lie_algebra [L' : lie_subalgebra R L] : lie_algebra R L' := {
+  bracket  := λ x y, ⟨⁅x.val, y.val⁆, L'.bracket x.property y.property⟩,
+  lie_add  := by { intros, apply set_coe.ext, apply lie_add, },
+  add_lie  := by { intros, apply set_coe.ext, apply add_lie, },
+  lie_self := by { intros, apply set_coe.ext, apply lie_self, },
+  jacobi   := by { intros, apply set_coe.ext, apply lie_ring.jacobi, },
+  lie_smul := by { intros, apply set_coe.ext, apply lie_smul, } }
+
+end lie_subalgebra
+
+section lie_module
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+variables (M  : Type v) [add_comm_group M] [module R M]
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/--
+A Lie module is a module over a commutative ring, together with a linear action of a Lie algebra
+on this module, such that the Lie bracket acts as the commutator of endomorphisms.
+-/
+class lie_module := (action : L →ₗ⁅R⁆ module.End R M)
```

Comment:
```suggestion
class lie_module :=
(action : L →ₗ⁅R⁆ module.End R M)
```
I'm not sure if this is the best definition. Usually, in mathlib, such definitions are factored out into "simpler" axioms. See, e.g. `mul_action` and `mul_action_distrib` etc... But this is just some feeling, on my side.
@digama0 Do you have a guideline here?

Reply:
Yep, I did notice my divergence from the pattern established by `mul_action` etc.

I was back and forth on the approach to take so I'm happy to fall in line with the idiom even if it's just based on a feeling.

I'll wait to here more before I mark this as resolved.

Reply:
Just to check I understand what you have in mind, is something approximately along the lines of the following preferable?
```
class add_action (α : Type u) (β : Type v)
  [add_monoid α] [has_zero β] [has_add β] extends has_scalar α β :=
(zero_smul : ∀ (b : β), (0 : α) • b = 0)
(add_smul : ∀ (x y : α) (b : β), (x + y) • b = x • b + y • b)

class has_linear_action (R : Type u) (L M : Type v)
  [ring R] [add_comm_group L] [add_comm_group M] [module R M] [module R L]
  extends add_action L M :=
(smul_add : ∀ (l : L) (m m' : M), l • (m + m') = l • m + l • m')
(smul_smul  : ∀ (r : R) (l : L) (m : M), (r • l) • m = r • (l • m))
(smul_smul' : ∀ (r : R) (l : L) (m : M), (r • l) • m = l • (r • m))

class has_lie_algebra_action (R : Type u) (L M : Type v)
  [comm_ring R] [add_comm_group L] [add_comm_group M] [lie_algebra R M] [lie_algebra R L]
  extends has_linear_action R L M :=
(bracket : ∀ (l l' : L) (m : M), ⁅l, l'⁆ • m = l • (l' • m) - l' • (l • m))
```
(And of course, this could be broken down even further.)

As I said, I find myself able to argue in favour of either approach. I can imagine that the above might have the convenience of requiring less coercion since the action is definitionally a function, but on the other hand there will presumably be a tonne of lemmas that need to be proved to make these classes convenient.

Perhaps you have other ideas?

All input from @jcommelin and @digama0 gratefully received!

Reply:
@ocfnash This was indeed what I had in mind (with `bracket` renamed to `lie_smul` :grinning:). I agree that your version is more concise. But it is also less granular. If you have some `linear_action` it can't be nicely promoted into a `lie_module` if it happens to be one. That's one reason why I'm in favour of the granular approach.
Also, concerning the "tonne of lemmas". I fear you'll need them anyway. With either approach, you'll have to write down 15 (or so) simp-lemmas etc, just to get the basics working smoothly.

Reply:
This is indeed what I had in mind (with `bracket` renamed to `lie_smul` :wink:). The benefit of the granular approach is that objects that already satisfy half of the axioms can gracefully be promoted to a full Lie module. With the 1-line definition, you'll probably make the write a bunch of compatibility glue. (I admit this is mostly intuition, no scientific arguments...)
Concerning the "tonne of lemmas". I guess that with both approaches you'll need to write some 15 (or so) simp-lemmas just to make the basics work smoothly.

Reply:
Thanks, I've had very little time this week but I'll put this in place over the weekend.

Reply:
I finally got round to this today, and have just pushed up two commits that aim to define Lie modules in a way closer to what you had in mind.

Annoyingly I was unable to keep the base `add_action` class or to extend `has_scalar` because `linear_action` and `lie_module` both depend on three carrier types and the would-be `add_action` and `has_scalar` depend on only two, which causes problems with the typeclass inference algorithm :-/

The [first commit](https://github.com/leanprover-community/mathlib/pull/1835/commits/0535c44eb284387d50f4175a4990ab5ed53ddd42) shows how I'd like to set things up whereas the [second commit](https://github.com/leanprover-community/mathlib/pull/1835/commits/0f43970204977b025fa1723d314450dc9ff80b96) is what I'm actually proposing.

I tried quite hard to do this well, and also experimented with keeping `add_action` (extending `has_scalar`) but making it external so that you need to supply an `add_action` instance to build a `linear_action` (rather than extending). This does work of course but things seemed to get very messy when I actually tried to use this setup.

Anyway, hopefully getting closer!